File: sdk/extension/testing/mock_transport.go
package exttesting

import (
	"context"
	"fmt"
	"io"
	"sync"

	"github.com/rodolfochicone/rc-project/sdk/extension"
)

// MockTransport is an in-memory message transport used by SDK tests.
type MockTransport struct {
	incoming   <-chan extension.Message
	outgoing   chan<- extension.Message
	localDone  chan struct{}
	remoteDone <-chan struct{}
	closeOnce  sync.Once
}

var _ extension.Transport = (*MockTransport)(nil)

// NewMockTransportPair creates a connected in-memory transport pair.
func NewMockTransportPair() (*MockTransport, *MockTransport) {
	return NewMockTransportPairWithBuffer(32)
}

// NewMockTransportPairWithBuffer creates a connected in-memory transport pair
// with the provided per-link buffer.
func NewMockTransportPairWithBuffer(buffer int) (*MockTransport, *MockTransport) {
	if buffer <= 0 {
		buffer = 1
	}

	leftToRight := make(chan extension.Message, buffer)
	rightToLeft := make(chan extension.Message, buffer)
	leftDone := make(chan struct{})
	rightDone := make(chan struct{})

	left := &MockTransport{
		incoming:   rightToLeft,
		outgoing:   leftToRight,
		localDone:  leftDone,
		remoteDone: rightDone,
	}
	right := &MockTransport{
		incoming:   leftToRight,
		outgoing:   rightToLeft,
		localDone:  rightDone,
		remoteDone: leftDone,
	}
	return left, right
}

// ReadMessage reads one queued message.
func (m *MockTransport) ReadMessage() (extension.Message, error) {
	if m == nil {
		return extension.Message{}, fmt.Errorf("read mock transport message: missing transport")
	}

	select {
	case <-m.localDone:
		return extension.Message{}, io.EOF
	case message := <-m.incoming:
		return message, nil
	}
}

// Receive reads one queued message with context cancellation.
func (m *MockTransport) Receive(ctx context.Context) (extension.Message, error) {
	if ctx == nil {
		return m.ReadMessage()
	}

	select {
	case <-ctx.Done():
		return extension.Message{}, context.Cause(ctx)
	case <-m.localDone:
		return extension.Message{}, io.EOF
	case message := <-m.incoming:
		return message, nil
	}
}

// WriteMessage enqueues one message for the peer transport.
func (m *MockTransport) WriteMessage(message extension.Message) error {
	if m == nil {
		return fmt.Errorf("write mock transport message: missing transport")
	}

	select {
	case <-m.localDone:
		return io.EOF
	case <-m.remoteDone:
		return io.EOF
	case m.outgoing <- message:
		return nil
	}
}

// Send enqueues one message with context cancellation.
func (m *MockTransport) Send(ctx context.Context, message extension.Message) error {
	if ctx == nil {
		return m.WriteMessage(message)
	}

	select {
	case <-ctx.Done():
		return context.Cause(ctx)
	case <-m.localDone:
		return io.EOF
	case <-m.remoteDone:
		return io.EOF
	case m.outgoing <- message:
		return nil
	}
}

// Close closes this transport endpoint.
func (m *MockTransport) Close() error {
	if m == nil {
		return nil
	}

	m.closeOnce.Do(func() {
		close(m.localDone)
	})
	return nil
}

File: sdk/extension-sdk-ts/src/testing/mock_transport.ts
import { EOFError, type Message, type Transport } from "../transport.js";

type PendingReader = {
  reject: (error: unknown) => void;
  resolve: (message: Message) => void;
};

/** An in-memory message transport used by SDK tests. */
export class MockTransport implements Transport {
  private readonly incoming: Message[] = [];
  private readonly readers: PendingReader[] = [];
  private peer?: MockTransport;
  private closed = false;
  private peerClosed = false;

  /** Connects this transport to its peer so messages written by one are readable by the other. */
  connect(peer: MockTransport): void {
    this.peer = peer;
  }

  /** Reads one queued message, blocking until a message is available. */
  async readMessage(): Promise<Message> {
    if (this.incoming.length > 0) {
      const message = this.incoming.shift();
      if (message === undefined) {
        throw new EOFError();
      }
      return message;
    }
    if (this.closed || this.peerClosed) {
      throw new EOFError();
    }
    return new Promise<Message>((resolve, reject) => {
      this.readers.push({ resolve, reject });
    });
  }

  /** Enqueues one message for the peer transport. */
  async writeMessage(message: Message): Promise<void> {
    if (this.closed || this.peerClosed || this.peer?.closed) {
      throw new EOFError();
    }
    this.peer?.pushIncoming(message);
  }

  /** Closes this transport endpoint. */
  async close(): Promise<void> {
    if (this.closed) {
      return;
    }
    this.closed = true;
    while (this.readers.length > 0) {
      this.readers.shift()?.reject(new EOFError());
    }
    this.peer?.markPeerClosed();
  }

  /** Reads one queued message. Alias for {@link readMessage}. */
  async receive(): Promise<Message> {
    return this.readMessage();
  }

  /** Enqueues one message for the peer transport. Alias for {@link writeMessage}. */
  async send(message: Message): Promise<void> {
    await this.writeMessage(message);
  }

  private pushIncoming(message: Message): void {
    const reader = this.readers.shift();
    if (reader !== undefined) {
      reader.resolve(message);
      return;
    }
    this.incoming.push(message);
  }

  private markPeerClosed(): void {
    this.peerClosed = true;
    while (this.readers.length > 0) {
      this.readers.shift()?.reject(new EOFError());
    }
  }
}

/** Creates a connected in-memory transport pair. */
export function createMockTransportPair(): [MockTransport, MockTransport] {
  const left = new MockTransport();
  const right = new MockTransport();
  left.connect(right);
  right.connect(left);
  return [left, right];
}
